content/blog/2019-09-25-interpreter-vector-support/index.md

```diff
@@ -0,0 +1,134 @@
+
+## Vector Instructions
+
+Single-Instruction Multiple-Data (SIMD) allows computer hardware to obtain signifigant speedups over the conventional execution paradigm, Multiple-Instruction Multiple-Data (MIMD). SIMD executes fewer instructions than MIMD to accomplish the same amount of work. In SIMD, multiple arithmetic operations are grouped together under one instruction. Generally, this aritmetic is executed over multiple execution units at the same instant (spatial SIMD), although some implementations allow the same instruction to use the same functional unit over multiple cycles (temporal SIMD). Various hardware architectures allow the programmer to support SIMD by exposing vector instructions in their ISA. Typically, there is a vector register file which holds multiple elements per register and can participate in vector arithmetic.
+
+## Vector Support in the Bril Interpreter
+
+### Memory Support
+
+Vector instructions are only useful when operating on large chunks of data. Conversely, registers are not designed to hold this amount of data. A prerequisite for vector instructions is support for data memory in the interpreter. The interpreter memory is implemented like a stack. The program is allowed to access any location in the fixed size memory. Stack frames were not implemented because the interpreter only supports a single function.
+
+Memory access requires both loads and store operations. To emulate arrays, we can loop over multiple addresses and perform a load and/or store at each location. Load and store instructions in Bril are implemented similarly to their assembly counterparts. A load takes in a memory address (an `int` register in Bril) and writes to a destination register. A store is an effect operation and does not write to a register. It uses two source registers: a register containing a value and a register containing an address. The Bril syntax is given below.
+
+```C
+// load value from address 2
+addr: int = const 2;
+data: int = lw addr;
+
+// store value to address 2
+sw data addr;
+```
+
+### Interpreted Vector Instructions
+
+Vector instructions were added to the Bril interpreter. We specifically implement fixed-sized vector instructions of length four (akin to native Intel `__m128` SSE instructions). Typescript and Javascript (Typescript always compiles to Javascript) do not have support for vector instrinsics in the current standard. Thus, we implement the Bril vector instructions as a loop over four values. Additionally, we add vector registers in Bril which must be used in the vector instructions. We target a vector-vector add (vvadd) program, so we include interpreter support for `vadd`, `vload`, and `vstore` instructions. The `vload` and `vstore` instructions communicate data between vector registers and the interpreter stack. The `vadd` instructions adds two vector registers and writes to a destination vector register. An example vvadd program is shown below.
```

Comment:
```suggestion
Vector instructions were added to the Bril interpreter. We specifically implement fixed-sized vector instructions of length four (akin to native Intel `__m128` SSE instructions). TypeScript and JavaScript (TypeScript always compiles to JavaScript) do not have support for vector instrinsics in the current standard. Thus, we implement the Bril vector instructions as a loop over four values. Additionally, we add vector registers in Bril which must be used in the vector instructions. We target a vector-vector add (vvadd) program, so we include interpreter support for `vadd`, `vload`, and `vstore` instructions. The `vload` and `vstore` instructions communicate data between vector registers and the interpreter stack. The `vadd` instructions adds two vector registers and writes to a destination vector register. An example vvadd program is shown below.
```

TS and JS both have two capital letters in them. (Please find/replace throughout.)